code/game/gamemodes/objective.dm

```diff
@@ -6,6 +6,46 @@
 	var/completed = 0					//currently only used for custom objectives.
 	var/dangerrating = 0				//How hard the objective is, essentially. Used for dishing out objectives and checking overall victory.
 	var/martyr_compatible = 0			//If the objective is compatible with martyr objective, i.e. if you can still do it while dead.
+	var/required_role = null			//If the objective can only be done by a certain role, such as absorb for changelings.
+	var/randomgen = 1
+
+/proc/generate_objectives(var/datum/mind/M, var/objectives_made = 2, var/has_escape = TRUE)
+	var/list/objective_types = (typesof(/datum/objective/default) - /datum/objective/default)
+	var/list/escape_objective_types = (typesof(/datum/objective/escape_obj) - /datum/objective/escape_obj)
+	for(var/datum/objective/O in objective_types)
+		if(O.required_role != M.special_role)
+			objective_types -= O
+		if(!O.randomgen)
+			objective_types -= O
+	for(var/i = 1, i <= objectives_made, i++)
+		if(i == objectives_made && has_escape)
+			var/datum/objective/holder_obj = pick(escape_objective_types)
+			var/datum/objective/O = new holder_obj
+			add_objective(M, O)
+			return
+		else
+			var/datum/objective/holder_obj = pick(objective_types)
+			var/datum/objective/O = new holder_obj
+			if(!O.martyr_compatible)
+				escape_objective_types -= list(/datum/objective/escape_obj/martyr)
+			add_objective(M, O)
+
+
+/proc/add_objective(var/datum/mind/M, var/datum/objective/O, var/announce_new_objectives = 0)
+	var/datum/objective/objective = new O
+	objective.owner = M
+	M.objectives += objective
+	objective.find_target()
+	objective.extra_prep()
+	objective.update_explanation_text()
+	if(announce_new_objectives)
+		if(M.special_role)
+			M.current << "<B>You are the [M.special_role].</B>"
+		var/obj_count = 1
+		for(var/datum/objective/OBJ in M.objectives)
+			M.current << "<B>Objective #[obj_count]</B>: [OBJ.explanation_text]"
+			obj_count++
+	return objective

```

Comment:
I don't see how the add_objective(M, O) lines above aren't going to cause runtimes because O is an already-instanced objective.